Make Iceberg Dynamic Filter test more resilient

alexjo2144 opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->

## Description

Fixes: https://github.com/trinodb/trino/issues/10932

<!-- Elaborate beyond the title of the PR as necessary to help the reviewers and maintainers.-->

## General information

Is this change a fix, improvement, new feature, refactoring, or other?

Flaky test fix

Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Test only

How would you describe this change to a non-technical end user or system administrator?

n/a
## Related issues, pull requests, and links

<!-- List any issue that is fixed and provide links to other related PRs, upstream release notes, and other useful resources:
* Fixes #issuenumber
* Related documentation in #issuenumber
* [Some release notes](http://usefulinfo.example.com)
-->

<!-- The following sections are filled in by the maintainer with input from the contributor:
Use :white_check_mark: or (x) or whatever really to signal selection.
-->

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

alexjo2144 commented:
I'll remove the `invocationCount` before merge, just wanted to test it a few times on CI

raunaqmorarka reviewed (commented):
lgtm % @findepi comments

losipiuk commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:
```diff
@@ -2732,26 +2732,29 @@ public void testAllAvailableTypes()
         long fullTableScan = (Long) computeActual("SELECT count(*) FROM lineitem").getOnlyValue();
-        long numberOfFiles = (Long) computeActual("SELECT count(*) FROM \"lineitem$files\"").getOnlyValue();
+        // Pick a value for totalprice where file level stats will not be able to filter out any data
+        MaterializedRow range = getOnlyElement(computeActual("SELECT max(lower_bounds[4]), min(upper_bounds[4]) FROM \"orders$files\"").getMaterializedRows());
```
I guess those are stats which correspond to `totalprice`

alexjo2144 replied:
Yes, 4 is the fieldId for totalPrice. I looked through the `$` metadata tables to see if there was a way to query for that value but couldn't find a good way of doing it

losipiuk commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:
```diff
@@ -2732,26 +2732,29 @@ public void testAllAvailableTypes()
+                "SELECT totalprice FROM orders WHERE totalprice > %s AND totalprice < %s LIMIT 1",
+                range.getField(0),
+                range.getField(1)))
+                .getOnlyValue();
```
But why are we sure gonna find it?
If we have two files with non overlaping ranges we get empty result range.

E.g: 
* f1: [10, 20]
* f2: [3,7]

We get range from the first query: [10, 7]. Which yields empty result from second query.

I am probably missing sth.

alexjo2144 replied:
That is possible, and would break the test, but seems unlikely. The overlapping range when I ran the test was most of the range of the column. Unless the data is explicitly ordered I think there should be some overlap.

If this still feels too flaky we can take it out and try testing the DF a different way.

alexjo2144 commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:
```diff
@@ -2732,26 +2732,29 @@ public void testAllAvailableTypes()
                 new QualifiedObjectName(ICEBERG_CATALOG, "tpch", "lineitem"));
 
-        // Assert no split level pruning occurs. If this starts failing a new totalprice may need to be selected
-        assertThat(probeStats.getTotalDrivers()).isEqualTo(numberOfFiles);
```
It is guaranteed by the query now, and driver count stats are kinda flaky.

aczajkowski replied:
I had same issue when adding MinIO test's for Iceberg. (this assertion was removed)
Would be nice to merge this PR earlier: https://github.com/trinodb/trino/pull/10894#discussion_r807733886
than https://github.com/trinodb/trino/pull/10894/checks

alexjo2144 commented:
I removed the `invocationCount` used for testing, and added another code comment about the assumption of there being an overlapping range in all files. @findepi @losipiuk
